src/mqttgw_ratelimitstate.erl

```diff
@@ -0,0 +1,90 @@
+-module(mqttgw_ratelimitstate).
+
+%% API
+-export([
+    new/0,
+    get_dirty/1,
+    get/2,
+    put_dirty/2,
+    put/4,
+    erase/0
+]).
+
+%% Types
+-type data() ::
+    #{time := non_neg_integer(),
+      bytes := non_neg_integer()}.
+-type result() ::
+    #{messages := [data()],
+      message_count := non_neg_integer(),
+      byte_count := non_neg_integer()}.
+-type constraints() ::
+    #{message_count := non_neg_integer(),
+      byte_count := non_neg_integer()}.
+
+-record(item, {
+    key :: mqttgw_id:agent_id(),
+    val :: [data()]
+}).
+
+-export_types([data/0, result/0]).
+
+%% Definitions
+-define(AGENT_TABLE, ?MODULE).
+
+%% =============================================================================
+%% API
+%% =============================================================================
+
+-spec new() -> ok.
+new() ->
+    catch _ = ets:new(
+        ?AGENT_TABLE,
+        [ set, public, named_table,
+          {keypos, #item.key} ]),
+    ok.
+
+-spec get_dirty(mqttgw_id:agent_id()) -> [data()].
+get_dirty(AgentId) ->
+    case ets:lookup(?AGENT_TABLE, AgentId) of
+        [#item{val=L}] -> L;
+        _ -> []
+    end.
+
+-spec get(mqttgw_id:agent_id(), non_neg_integer()) -> result().
+get(AgentId, Time) ->
+    L = [Data || #{time := T} = Data <- get_dirty(AgentId), T > Time],
+    BytesCount = lists:foldl(fun(#{bytes := B}, Acc) -> Acc + B end, 0, L),
+    MessageCount = length(L),
+    #{messages => L,
+      message_count => MessageCount,
+      byte_count => BytesCount}.
+
+-spec put_dirty(mqttgw_id:agent_id(), data()) -> ok.
+put_dirty(AgentId, Data) ->
+    L = get_dirty(AgentId),
+    Item = #item{key=AgentId, val=[Data|L]},
+    ets:insert(?AGENT_TABLE, Item),
+    ok.
+
+-spec put(mqttgw_id:agent_id(), data(), non_neg_integer(), constraints()) -> ok | {error, atom()}.
+put(AgentId, Data, Time, Constraints) ->
+    #{bytes := Bytes} = Data,
+    #{message_count := MessageCountLimit,
+      byte_count := ByteCountLimit} = Constraints,
+    Result = get(AgentId, Time),
+    case Result of
+        #{message_count := MessageCount} when (MessageCount + 1) > MessageCountLimit ->
```

Comment:
`MessageCount >= MessageCountLimit`?

Reply:
Why?

Reply:
+1 is not obvious if only reason to do it is using `>` instead of `>=`

Reply:
In my opinion, what is more important is to follow the same approach:
```erlang
(MessageCount + 1) > MessageCountLimit
(ByteCount + Bytes) > ByteCountLimit
```

Reply:
ok lets leave it as is